query/query.go

```diff
@@ -234,24 +238,31 @@ func NewResultBuilder(q Query) *ResultBuilder {
 	b := &ResultBuilder{
 		Query:  q,
 		Output: make(chan Result, bufSize),
+		closed: make(chan struct{}),
 	}
-	b.Process = goprocess.WithTeardown(func() error {
+
+	b.ctx, b.cancel = context.WithCancel(context.Background())
+	context.AfterFunc(b.ctx, func() {
+		b.wg.Wait()
 		close(b.Output)
-		return nil
+		close(b.closed)
 	})
 	return b
 }
 
 // ResultsWithChan returns a Results object from a channel
 // of Result entries.
 //
-// DEPRECATED: This iterator is impossible to cancel correctly. Canceling it
-// will leave anything trying to write to the result channel hanging.
+// DEPRECATED: This iterator takes sepcial care to cancel correctly. Canceling
```

Comment:
Good chance to get rid of this method...

Reply:
I could get rid of it... or leave it now that it is technically fixed because callers can now select over `Results.Done` when writing to the results input channel.